diff --git a/cipher/poly1305.c b/cipher/poly1305.c
index e57e64f..9c26bc9 100644
--- a/cipher/poly1305.c
+++ b/cipher/poly1305.c
@@ -289,6 +289,8 @@ static unsigned int poly1305_final (poly1305_context_t *ctx,
 
 #ifdef USE_MPI_32BIT
 
+#undef HAVE_COMPATIBLE_GCC_ARM_PLATFORM_AS
+
 #ifdef HAVE_COMPATIBLE_GCC_ARM_PLATFORM_AS
 
 /* HI:LO += A * B (arm) */
diff --git a/mpi/ec-inline.h b/mpi/ec-inline.h
index a07826e..cfe380a 100644
--- a/mpi/ec-inline.h
+++ b/mpi/ec-inline.h
@@ -755,7 +755,7 @@ LIMB64_HILO(mpi_limb_t hi, mpi_limb_t lo)
 
 #endif /* __i386__ */
 
-
+#undef HAVE_COMPATIBLE_GCC_ARM_PLATFORM_AS
 /* ARM addition/subtraction helpers.  */
 #ifdef HAVE_COMPATIBLE_GCC_ARM_PLATFORM_AS
 
