diff --git a/build/cmake/CMakeModules/AddZstdCompilationFlags.cmake b/build/cmake/CMakeModules/AddZstdCompilationFlags.cmake
index e23b9d60..0f9b5be2 100644
--- a/build/cmake/CMakeModules/AddZstdCompilationFlags.cmake
+++ b/build/cmake/CMakeModules/AddZstdCompilationFlags.cmake
@@ -56,6 +56,10 @@ macro(ADD_ZSTD_COMPILATION_FLAGS)
         if (CMAKE_BUILD_TYPE MATCHES "Debug")
             EnableCompilerFlag("/DDEBUGLEVEL=1" true true)
         endif ()
+
+        EnableCompilerFlag("/sdl" true true)
+        EnableCompilerFlag("/guard:cf" true true)
+
     endif ()
 
     # Remove duplicates compilation flags
